src/input_streams.rs

```diff
@@ -375,28 +375,49 @@ impl<'a> InputStreams<'a> {
                 left,
                 right,
             }) => {
-                let x = self.input_value(&UserInput::Single(*right)).abs()
-                    - self.input_value(&UserInput::Single(*left)).abs();
-                let y = self.input_value(&UserInput::Single(*up)).abs()
-                    - self.input_value(&UserInput::Single(*down)).abs();
+                let x = self.input_value(&UserInput::Single(*right), true).abs()
+                    - self.input_value(&UserInput::Single(*left), true).abs();
+                let y = self.input_value(&UserInput::Single(*up), true).abs()
+                    - self.input_value(&UserInput::Single(*down), true).abs();
                 Some(DualAxisData::new(x, y))
             }
             _ => None,
         }
     }
 
     fn extract_dual_axis_data(&self, dual_axis: &DualAxis) -> DualAxisData {
-        let x = self.input_value(&UserInput::Single(InputKind::SingleAxis(dual_axis.x)));
-        let y = self.input_value(&UserInput::Single(InputKind::SingleAxis(dual_axis.y)));
-
-        if x > dual_axis.x.positive_low
-            || x < dual_axis.x.negative_low
-            || y > dual_axis.y.positive_low
-            || y < dual_axis.y.negative_low
-        {
-            DualAxisData::new(x, y)
-        } else {
-            DualAxisData::new(0.0, 0.0)
+        match dual_axis.deadzone_shape {
+            DeadzoneShape::Cross => {

```

Comment:
This logic should all be refactored into a reusable public method or two on `DeadZoneShape` :)